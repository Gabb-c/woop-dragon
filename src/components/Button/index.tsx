import styles from './styles.module.scss';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string;
  loading?: boolean;
  flavor?: 'delete' | 'save' | 'default';
}

export const Button: React.FC<ButtonProps> = ({
  label,
  flavor,
  loading,
  ...props
}): JSX.Element => {
  return (
    <button
      {...props}
      className={!flavor || flavor === 'default' ? styles.button : styles[`button-${flavor}`]}
    >
      {!loading ? label : 'loading...'}
    </button>
  );
};
