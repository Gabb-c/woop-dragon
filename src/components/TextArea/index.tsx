import styles from './styles.module.scss';

interface TextAreaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  label?: string;
  error?: boolean;
  errorText?: string;
}

export const TextArea: React.FC<TextAreaProps> = ({
  label,
  error,
  errorText,
  ...props
}): JSX.Element => {
  return (
    <div className={styles.container}>
      <span className={error ? styles.inputErrorLabel : styles.inputLabel}>{label}</span>
      <textarea {...props} data-testid="textarea-test" className={styles.input} />
      {error && <span className={styles.inputErrorText}>{errorText}</span>}
    </div>
  );
};
