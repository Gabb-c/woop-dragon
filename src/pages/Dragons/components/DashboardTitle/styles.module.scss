.container {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  flex-direction: row;

  gap: 1.6em;
  margin-bottom: 2em;

  color: #fefefe;

  transition: all 0.6s ease;

  h1 {
    line-height: 1.2em;
  }
}
