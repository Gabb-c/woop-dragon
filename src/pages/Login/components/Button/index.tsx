import styles from './styles.module.scss';
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  label?: string;
  loading?: boolean;
  flavor?: 'delete' | 'edit' | 'default';
}

export const Button: React.FC<ButtonProps> = ({ label, loading, ...props }): JSX.Element => {
  return (
    <button {...props} className={styles.button}>
      {!loading ? label : 'loading...'}
    </button>
  );
};
