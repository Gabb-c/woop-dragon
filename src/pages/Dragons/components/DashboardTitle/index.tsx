import styles from './styles.module.scss';
import { Button } from '@components/Button';
import { useAuth } from '@contexts/AuthContext';
import { useNavigate } from 'react-router-dom';

export const DashboardHeader = () => {
  const { signOut } = useAuth();

  const navigate = useNavigate();

  return (
    <div className={styles.container}>
      <Button label="Logout" flavor="delete" onClick={() => signOut()} />
      <h1>Dragons Dashboard</h1>
      <Button label="Add" flavor="save" onClick={() => navigate('/dragons/form')} />
    </div>
  );
};
