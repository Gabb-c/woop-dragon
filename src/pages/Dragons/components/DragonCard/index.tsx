import { Dragon } from '@models/dragon';
import { format } from 'date-fns';
import { Button } from 'src/pages/Login/components/Button';
import styles from './styles.module.scss';

interface DragonCardProps {
  dragon: Dragon;
  onDelete: (id: string) => void;
  onEdit?: (id: string) => void;
}

export const DragonCard: React.FC<DragonCardProps> = ({ dragon, onDelete }): JSX.Element => {
  return (
    <div className={styles.card}>
      <h5 className={styles.dragonType}>{dragon.type.toUpperCase()}</h5>
      <h1 className={styles.dragonName}>{dragon.name}</h1>
      <h3 className={styles.dragonBirth}>{format(new Date(dragon.createdAt), 'MM/dd/yyyy')}</h3>
      <div className={styles.fields}>
        <Button label="Edit" />
        <Button label="Delete" onClick={() => onDelete(dragon.id)} />
      </div>
    </div>
  );
};
