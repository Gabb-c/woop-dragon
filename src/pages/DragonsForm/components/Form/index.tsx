import DragonService from '@services/dragon';
import styles from './styles.module.scss';
import { Button } from '@components/Button';
import { Dragon } from '@models/dragon';
import { PropagateLoader } from 'react-spinners';
import { TextArea } from '@components/TextArea';
import { TextField } from '@components/TextField';
import { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';

export const Form: React.FC = (): JSX.Element => {
  const navigate = useNavigate();
  const { id } = useParams();

  const [dragon, setDragon] = useState<Dragon>({} as Dragon);
  const [loading, setLoading] = useState(false);
  const [loadingButton, setLoadingButton] = useState(false);

  const handleLoadDragon = async () => {
    if (!id) return;
    setLoading(true);
    await DragonService.getDragonById(id)
      .then((response) => {
        setDragon(response);
        return setLoading(false);
      })
      .catch(() => navigate('/dragons'));
  };

  const handleSubmit = () => {
    setLoadingButton(true);
    if (!id) {
      DragonService.createDragon(dragon)
        .then(() => navigate('/dragons'))
        .catch((_error) => console.error(_error));
    } else {
      DragonService.editDragon(dragon, dragon.id)
        .then(() => navigate('/dragons'))
        .catch((_error) => console.error(_error));
    }
  };

  useEffect(() => {
    handleLoadDragon().catch((_error) => console.error(_error));
  }, []);

  const { name, histories, type } = dragon;

  return (
    <div className={styles.card}>
      <h2>{id ? `Edit Dragon #${id}` : 'Add new Dragon'}</h2>
      <div className={styles.fields}>
        {loading ? (
          <PropagateLoader color="white" />
        ) : (
          <>
            <TextField
              label="Name"
              name="name"
              value={name}
              onChange={({ target: { value } }) => {
                setDragon({ ...dragon, name: value });
              }}
            />
            <TextField
              label="Type"
              name="type"
              value={type}
              onChange={({ target: { value } }) => {
                setDragon({ ...dragon, type: value });
              }}
            />
            <TextArea
              label="Histories"
              name="histories"
              value={histories}
              maxLength={200}
              rows={3}
              onChange={({ target: { value } }) => {
                setDragon({ ...dragon, histories: value });
              }}
            />
            <Button
              label="Save"
              name="save-dragon"
              flavor="save"
              loading={loadingButton}
              disabled={loading}
              onClick={handleSubmit}
            />
            <Button
              label="Cancel"
              name="cancel"
              type="submit"
              flavor="delete"
              onClick={() => navigate('/dragons')}
            />
          </>
        )}
      </div>
    </div>
  );
};
