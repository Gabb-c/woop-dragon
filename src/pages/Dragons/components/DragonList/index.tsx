import { useEffect, useState } from 'react';
import { PropagateLoader } from 'react-spinners';

import DragonService from '@services/dragon';

import { DragonCard } from '../DragonCard';
import { Dragon } from '@models/dragon';

import styles from './styles.module.scss';
import { DashboardHeader } from '../DashboardTitle';

export const DragonList: React.FC = (): JSX.Element => {
  const [list, setList] = useState<Dragon[]>([]);
  const [loading, setLoading] = useState(true);

  const loadDragons = async () => {
    setLoading(true);
    await DragonService.getAllDragons()
      .then((response) => {
        setList(response);
        return setLoading(false);
      })
      .catch((error) => console.error(error));
  };

  const handleDelete = (id: string) => {
    setLoading(true);
    DragonService.deleteDragonById(id)
      .then(() => {
        return loadDragons();
      })
      .catch((error) => console.error(error));
  };

  useEffect(() => {
    loadDragons().catch((error) => console.error(error));
  }, []);

  return (
    <div className={styles.container}>
      <DashboardHeader />
      {loading ? (
        <PropagateLoader color="white" />
      ) : (
        <div className={styles.lisCtontainer}>
          {list.map((d, i) => (
            <DragonCard onDelete={handleDelete} dragon={d} key={i} />
          ))}
        </div>
      )}
    </div>
  );
};
