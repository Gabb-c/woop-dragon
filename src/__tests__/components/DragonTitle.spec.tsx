import React from 'react';
import { render, screen } from '@testing-library/react';
import { DashboardHeader } from '../../pages/Dragons/components/DashboardTitle';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';

describe('Dragon Title', () => {
  it('test if it renders in the DOM', () => {
    render(<DashboardHeader />, { wrapper: BrowserRouter });

    const resTest = screen.getByText(/dragons dashboard/i);

    expect(resTest).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    render(<DashboardHeader />, { wrapper: BrowserRouter });

    const resTest = screen.getByText(/dragons dashboard/i);

    expect(resTest).toMatchSnapshot();
  });
});
