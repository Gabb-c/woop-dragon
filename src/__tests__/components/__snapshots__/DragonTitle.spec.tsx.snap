// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragon Title should match snapshot 1`] = `
<h1>
  Dragons Dashboard
</h1>
`;
